### ISSUE TITLE
Applying an empty farms list is a no-op

### ISSUE THREAD

**Issue by ghost** [zevenet/nftlb#12]
Hi. Both of the following JSON configurations are accepted by nftlb without error, both when launched and if submitted by way of a POST.

```json
{}
{ "farms": [] }
```

However, nftlb does nothing in response. If nftlb is launched with this config, it does not create an "nftlb" table. In and as of itself, that isn't necessarily a problem. However, it becomes a problem if the table had been created and populated previously, because nftlb doesn't do any clean-up work.

To reproduce:-

1. Launch nftlb with a populated farms list (I used the same config as mentioned in issue #11)
2. Observe that the nftlb table and applicable chains have been created
3. Terminate nftlb
4. Launch nftlb with one of the empty configurations shown above
5. Observe that step 4 changes nothing and that the prior tables/chains remain in place

The same applies where using the JSON API for step 4, even though the response code is "success". It might seem strange to try to use an 'empty' configuration, but I think it has some utility, especially where the farms list is explicitly specified as an empty list. In particular, I'm hoping to ship an empty config in the distro package that I'm creating, albeit with a recommendation to install one of the bundled sample configs.

In summary, I think that it would be useful for nftlb to act upon an empty farms list. If that is not possible, then it woud be helpful for it to raise an error i.e. printing something to STDERR when starting and/or returning a failure code when using the API. For now, I am working around these pitfalls by adding additional post-start sanity checks in the OpenRC runscript that I am working on.

**Comment by ghost** [comment]
Further testing indicates that the issue isn't just limited to the two configurations shown in the previous comment. It also applies to almost-correct configurations, such as the one below. Note the accidental specification of "ip6" as the value for the "family" key instead of "ipv6". As before, nftlb does not attempt to apply the configuration. However, it does accept the configuration without raising any sort of error. That makes it difficult for the user to notice that anything is wrong. Once the user does realise that there is a problem, the absence of an error code/message then makes it difficult to diagnose where the mistake was made.

By the way, it also silently fails to apply the config if the value for "family" is "dual".

```json
{
  "farms": [
    {
      "name": "lb01",
      "family": "ip6",
      "virtual-addr": "fe80::21b:21ff:fecf:5f9",
      "virtual-ports": "443,6595",
      "mode": "dnat",
      "protocol": "tcp",
      "scheduler": "weight",
      "state": "up",
      "backends": [
        {
          "name": "gs1",
          "ip-addr": "fe80::5054:ff:fe1d:cc48",
          "port": "6595",
          "weight": "500",
          "mark": "0x1",
          "state": "up"
        }
      ]
    }
  ]
}
```

**Comment by nevola** [comment]
Hi @kerframil 

In regards to the first report, we could add an OK message though no action has been applied.

In regards to the second message, the input validation will show via syslog that the parameter is not valid and the default (ipv4) will be used. We can change that in order to send this message in the API response.

**Comment by ghost** [comment]
> In regards to the first report, we could add an OK message though no action has been applied.

Being able to discern between those two outcomes could be helpful.

> In regards to the second message, the input validation will show via syslog that the parameter is not valid and the default (ipv4) will be used.

That doesn't seem to be happening here. Instead, the error is raised by nft(8) itself, with nftlb then writing its STDERR to syslog with the user facility. The error message is _extremely_ long and difficult to read. It seems that the design of nft's error messages is heavily skewed towards interactive usage in a tty (it would be nice to reduce the verbosity here). If nftlb itself is detecting the bad value of "ip6", it is not currently reporting the fact. At least, not for the config sample shown. **EDIT:** Even with a log level of 7.

> We can change that in order to send this message in the API response.

That would be great! The client would then have a way to unambiguously determine that an error occurred and have the option of propagating an error message to the user, if provided.

**Comment by nevola** [comment]
> That doesn't seem to be happening here. Instead, the error is raised by nft(8) itself, with nftlb then writing its STDERR to syslog with the user facility. The error message is _extremely_ long and difficult to read. It seems that the design of nft's error messages is heavily skewed towards interactive usage in a tty (it would be nice to reduce the verbosity here). If nftlb itself is detecting the bad value of "ip6", it is not currently reporting the fact. At least, not for the config sample shown. **EDIT:** Even with a log level of 7.

It should be raised without any debug. The message would be something like:

"parsing unknown value XXXX, using defaults"

**Comment by ghost** [comment]
> It should be raised without any debug. The message would be something like:
> 
> "parsing unknown value XXXX, using defaults"

Apologies. I see it now. It was hard to spot admist the huge error messages from nft(8), which also include a lot of horizontal and vertical whitespace that gets logged.

**Comment by nevola** [comment]
> It was hard to spot admist the huge error messages from nft(8), which also include a lot of horizontal and vertical whitespace that gets logged.

That should be nft command errors. Some command maybe have left inconsistencies in the objects, if you provide some steps to reproduce it I could give it a hint.

**Comment by ghost** [comment]
Hmm. The parser doesn't complain about the bad value unless the syslog level is at least 6 (I normally use 5). That explains why I did not see it initially.

My steps were to ensure that no nftlb table exists, then spawn nftlb with the bad "ip6" config shown in a previous comment. Here are the results for me at log levels 5 through 7.

- [nftlb-loglevel-5.txt](https://gist.githubusercontent.com/kerframil/fd1f9944f5f024360a354cb5a493a3fc/raw/5eec4f2387652f2835e0fd316f2ebebd406657fb/nftlb-loglevel-5.txt)
- [nftlb-loglevel-6.txt](https://gist.githubusercontent.com/kerframil/c3e5257cf7b0e3d0bd0ef618826a88d3/raw/a85a4af82d51c31041a079ddc506fffd5c86db54/nftlb-loglevel-6.txt)
- [nftlb-loglevel-7.txt](https://gist.githubusercontent.com/kerframil/026395be7f6fd009e9d6dcc64a764016/raw/caef2e868d70858dd3b8b0eaca6bb1c21884eded/nftlb-loglevel-7.txt)

Wearing my sysadmin cap, these are my thoughts (just my proverbial two cents):-

- Though it is not the fault of nftlb, the nft output is really badly formatted and messy (the browser makes it look better than it really is, try downloading and viewing with less).
- As a user, I would rather not see the raw nft STDERR output other than at level 7. At least, I don't think it belongs in level 5. For 5, I feel that a message of severity "err" which states only that nft finished with whichever (non-zero) exit status value would be enough.
- I am most interested in warnings or errors that are produced by nftlb itself. Especially config parsing errors! Personally, I would consider an unrecognised/bad value as an error (putting it well above "info" level).
- I would actually prefer that nftlb does not try to guess the intent if a bad value is provided but, instead, unambiguously complain and raise an error message before bailing out. I would even prefer for nftlb to exit if spawned with a bad config, as many daemons do. Also, for the JSON API to be more strident about reporting errors, as discussed.

Regarding the last point, this is really just about having the behaviour be predictable and to be informed when something isn't right, rather than nftlb just trying to carry on and maybe (or maybe not) doing what the user intended.

**Comment by ghost** [comment]
I ended up working around this issue by having my "reload" service action [perform a DELETE](https://bitbucket.org/kerframil/portage-overlay/src/472060eecb472f9e7bfb6c2938b93028fe2396ca/net-misc/nftlb/files/nftlb.initd#lines-172) before POSTing, as was recommended. Along with the modification to nft_reset, it seems to be working well.

I'm now also using jshon to check whether the supplied configuration contains an empty farms list before submitting to nftlb. If not empty, a check is conducted post-reload to ensure that at least one "nftlb" table exists. For now, I have found that this is necessary to advise the user of situations where the config is not accepted and acted upon by nftlb, even though "success" is reported.

**Comment by nevola** [comment]
> I ended up working around this issue by having my "reload" service action [perform a DELETE](https://bitbucket.org/kerframil/portage-overlay/src/472060eecb472f9e7bfb6c2938b93028fe2396ca/net-misc/nftlb/files/nftlb.initd#lines-172) before POSTing, as was recommended. Along with the modification to nft_reset, it seems to be working well.
> 

Yes, I believe that is the best option.

> I'm now also using jshon to check whether the supplied configuration contains an empty farms list before submitting to nftlb. If not empty, a check is conducted post-reload to ensure that at least one "nftlb" table exists. For now, I have found that this is necessary to advise the user of situations where the config is not accepted and acted upon by nftlb, even though "success" is reported.

Is this json validation useful to be done at nftlb side?

**Comment by ghost** [comment]
> > I'm now also using jshon to check whether the supplied configuration contains an empty farms list before submitting to nftlb. If not empty, a check is conducted post-reload to ensure that at least one "nftlb" table exists. For now, I have found that this is necessary to advise the user of situations where the config is not accepted and acted upon by nftlb, even though "success" is reported.
> 
> Is this json validation useful to be done at nftlb side?

Oh, definitely. The API is the conduit by which other programs are expected to control to nftlb. As with any API, its consumers depend on the provider to state whether an operation succeeds or fails and, ideally, why. At present, a status code of 2xx and a body of ```{ "response" : "success" }``` only seems to guarantee one thing: that nftlb's http server was able to decode the request body. It cannot be relied upon to indicate whether nftlb took action and whether it did so successfully. If something goes wrong, it cannot be relied upon to convey a basic error mesage explaining why.

That's why I have put more effort into implementing auxiliary sanity checks in the runscript than I would for most daemons. But no exernal program is privy to the exact internal state of nftlb; the best source of diagnostic information concerning nftlb can only truly be nftlb itself. Taking my check_tables function as an example, all it can do is meekly point out that something might be wrong after a start/reload action and suggest that the user check syslog. Even then, the log level threshold needs to be at least "6", otherwise messages that are important for the user to be able to troubleshoot config issues may be missing.

Also, I have just finished updating a program that my employer uses to handle weight changes (I'm not yet using nftlb in production but hope to be next week). It works quite well. However, once the changes have been POSTed, I need to issue another GET request and verify that all the changes were applied as expected. It's not an onerous requirement but, if the API had better error reporting, I probably would not have needed to write this additional verification code.

TL;DR: After launching nftlb with a configuration - or POSTing a configuration to it - I have no idea whether nftlb really did what was intended without applying crude heuristics or human oversight. It is a great tool but, if there is one thing that could make it greater still, I would say that better error reporting at the API level would be it.

**Comment by nevola** [comment]
Yes, I agree. As I've been mainly focused on the load balancing features, the API is too basic. But thank you for your feedback, I'll work on that very soon.

**Comment by nevola** [comment]
Commit applied, reopen if need any rework.

**Comment by ghost** [comment]
Thank you very much. I just deployed nftlb today, as it happens. I will test this change (and the others) on the inactive balancer soon.

**Comment by nevola** [comment]
Great, thank you. Please note that these changes are in the devel branch.
